### ISSUE TITLE
Serialization: Table columns with "Display values"="Use foreign key" no set on `load` depending on creation order

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#15848]
**Describe the bug**
When a table column is using "Display values"="Use foreign key", then that information is not set on load if the referenced column isn't already created.
This is caused by of the way serialization loads objects in alphabetic form, and doesn't retry - very similar to #14912.

**To Reproduce**
1. Setup a instance, change the name of Sample Dataset database, so it's easier to distinguish - let's call it "SOURCE"
2. Admin > Data Model > "SOURCE" > Orders set "Use foreign key" for User_ID and Product_ID, and for Reviews do the same for Product_ID
2. Serialization `dump` - notice that it's creating yaml files correctly in `fks` directory
3. Serialization `load` - no errors shown.
4. On target: Go to Admin > Data Model > "SOURCE" > Orders - both User_ID and Product_ID will have "Use original value", while Reviews.Product_ID will correctly have "Use foreign key".

**Information about your Metabase Installation:**
Tested 1.37.9 thru 1.39.1 and `995f444`, which fixes #14912

**Severity**
Currently only giving it P3

**Comment by camsaul** [comment]
Seems like a key not getting carried over should be an easy fix?

**Comment by flamber** [comment]
@camsaul No, it depends on if the object has been created or not. Same as many of the question issues, which was handled by adding the objects to a queue and then re-running after the dependencies has been created.

**Comment by jeff303** [comment]
@flamber , what you're referring to is actually part of the dimensions, if I'm not mistaken?  Ex `dimensions.yaml`:

```
- field_id: /databases/SOURCE/schemas/public/tables/orders/fks/user_id
  name: User ID
  type: external
  human_readable_field_id: /databases/SOURCE/schemas/public/tables/people/fields/id
- field_id: /databases/SOURCE/schemas/public/tables/orders/fks/product_id
  name: Product ID
  type: external
  human_readable_field_id: /databases/SOURCE/schemas/public/tables/products/fields/title
  ```
  
  Is that the problem, or is the problem with `fk_target_field_id` under `/tables/orders/fks/product_id.yaml` (for instance)?

**Comment by flamber** [comment]
@jeff303 I can still reproduce on 0.40.1 - I'm unsure if this is caused by an ordering issue or race-condition, since you cannot reproduce. But my `dimensions.yaml` (I didn't know this file controlled remapping, thought it was in the fks directories) only contains this, so it's not a problem during load as I initially thought, but during dump:
```
- field_id: /databases/SOURCE-Sample Dataset/schemas/PUBLIC/tables/REVIEWS/fks/PRODUCT_ID
  name: Product ID
  type: external
  human_readable_field_id: /databases/SOURCE-Sample Dataset/schemas/PUBLIC/tables/PRODUCTS/fields/TITLE
```
And the database has remapping for Orders - and it works as expected (on source, before dump and load into a target):
![image](https://user-images.githubusercontent.com/1447303/127150089-9091f5ae-6eda-427d-abff-211ad30235f7.png)
I ran with tracing on `metabase-enterprise.serialization` and got absolutely no extra information compared to normally.
```
java  -Dlog4j.configurationFile="https://log4j.us/v2/templates/metabase?trace=metabase-enterprise.serialization" -DMB_DB_FILE=source.db -DMB_JETTY_PORT=3000 -jar metabase-1.40.1.jar dump source;
...
2021-07-27 14:03:43,947 INFO serialization.cmd :: BEGIN DUMP to source via user null
2021-07-27 14:03:44,352 INFO serialization.cmd :: END DUMP to source via user null
```

**Comment by jeff303** [comment]
> @jeff303 I can still reproduce on 0.40.1 - I'm unsure if this is caused by an ordering issue or race-condition, since you cannot reproduce. But my `dimensions.yaml` (I didn't know this file controlled remapping, thought it was in the fks directories) only contains this, so it's not a problem during load as I initially thought, but during dump:
> 
> ```
> - field_id: /databases/SOURCE-Sample Dataset/schemas/PUBLIC/tables/REVIEWS/fks/PRODUCT_ID
>   name: Product ID
>   type: external
>   human_readable_field_id: /databases/SOURCE-Sample Dataset/schemas/PUBLIC/tables/PRODUCTS/fields/TITLE
> ```


So to be absolutely clear, the problem is that the `User ID` field wasn't showing up in your `dimensions.yaml` file on dump, even though it apparently was on mine?

**Comment by flamber** [comment]
@jeff303 What I posted was the full dimensions.yaml even though I defined remapping on Orders.Product_ID, Orders.User_ID and Reviews.Products_ID

**Comment by piranha** [comment]
It works properly now and there is a test about it: [dimensions-test](https://github.com/metabase/metabase/blob/master/enterprise/backend/test/metabase_enterprise/serialization/v2/extract_test.clj#L647-L650).
